xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/doc/XWikiAttachmentList.java

```diff
@@ -0,0 +1,243 @@
+    {
+        return addAll(c);
+    }
+
+    @Override
+    public XWikiAttachment remove(int index)
+    {
+        XWikiAttachment removedAttachment = map.remove(this.decorated().get(index).getFilename());
+        if (removedAttachment != null) {
+            updatedMap();
+        }
+        return removedAttachment;
+    }
+
+    /**
+     * Removes XWikiAttachment.
+     * 
+     * @param attachment XWikiAttachment to remove.
+     * @return true unless the attachment is not found
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean remove(Object attachment)
+    {
+        XWikiAttachment xwikiAttachment = (XWikiAttachment) attachment;
+        if (map.remove(xwikiAttachment.getFilename()) == null) {
+            return false;
+        }
+        updatedMap();
+        return true;
+    }
+
+    /**
+     * Adds or replaces attachment with the same filename as the parameter.
+     * 
+     * @param attachment the attachment to add to the list
+     * @return the attachment that was previously matched to the same filename or null if no attachment was matched to
+     *         it
+     * @since 10.0RC1
+     */
+    public XWikiAttachment set(XWikiAttachment attachment)
+    {
+        XWikiAttachment put = map.put(attachment.getFilename(), attachment);
+        if (put == null || !put.equals(attachment)) {
+            added(attachment);
+            updatedMap();
+        }
+        return put;
+    }
+
+    /**
+     * Adds or replaces attachment with the same filename as the parameter.
+     * 
+     * @param index this parameter is not used but is needed to override the method
+     * @param attachment the attachment to add to the list
+     * @return the attachment that was previously matched to the same filename or null if no attachment was matched to
+     *         it
+     * @since 10.0RC1
+     */
+    @Override
+    public XWikiAttachment set(int index, XWikiAttachment attachment)
+    {
+        return set(attachment);
+    }
+
+    /**
+     * @param filename the filename of the attachment to be returned.
+     * @return attachment with the given filename or null if not found.
+     * @since 10.0RC1
+     */
+    public XWikiAttachment getByFilename(String filename)
+    {
+        return map.get(filename);
+    }
+
+    @Override
+    public boolean removeAll(Collection<?> c)
+    {
+        boolean changed = false;
+        XWikiAttachmentList list = (XWikiAttachmentList) this.decorated();
+        for (XWikiAttachment x : (Collection<? extends XWikiAttachment>) c) {
+            if (list.contains(x)) {
+                if (remove(x)) {
+                    changed = true;
+                }
+            }
+        }
+        return changed;
+    }
+
+    @Override
+    public boolean retainAll(Collection<?> c)
+    {
+        boolean changed = false;
+        XWikiAttachmentList list = (XWikiAttachmentList) this.decorated();
+        for (XWikiAttachment x : list) {
+            if (remove(x)) {
```

Comment:
Actually you are supposed to remove the elements which are NOT in the input list.